** Introduction
Business Context: Transportation systems are complex and interesting for a variety of reasons.... Having obtained multiple open data sets published by the Toronto Transit Commission, it would be interesting to extract relevant insights from it. Due to the size and format of these datasets, using Excel alone will prove to be a difficult task especially as it has multiple variables. 

Business Problem: My task here is to understand the information that is available and extract key insights. Specifically, I want to answer the following six questions:
1. What is the average delay across buses plying the same route?
2. What is the average delay across the ttc bus system?
3. What is the most frequent incident type that leads to bus delays?
4. What is the time period at which most delays occur?
5. What is the route with the highest average delay?
6. What is the time period which the highest average delay times?

Analytical Context: The data is found from Toronto's open data website and contains infromation on bus delays by routes and year. These datasets of delays related to ttc buses are from 2014 to 2017 and are refreshed monthly. It was last refreshed on July 20, 2023. I will: 1) manipulate the data to make it suitable for analysis; and 2) find the relevant players in the different categories related ot the research questions. 

* ISSUES IDENTIFIED WITH RAW DATASET
* From years 2014 to 2021, data is recorded monthly in different worksheets for each workbook representing a specific year. This is different for the years 2022 and 2023.


################################################## DATA WRANGLING #########################################################

In [1]:
%pip install openpyxl
import pandas as pd
import numpy as np


[notice] A new release of pip available: 22.2.2 -> 23.2.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


Datasets from 2014 to 2021, have 12 excel worksheets, one for each month. For the year 2022 and 2023, the data have been compiled into one worksheet. Using the pd.read_excel and pd.concat functions, bus data for the following years are merged into one worksheet for each year to facilitate data manipulation. 

In [12]:
yr_2014 = pd.concat(pd.read_excel("ttc-bus-delay-data-2014.xlsx",sheet_name = None),ignore_index = True)
yr_2015 = pd.concat(pd.read_excel("ttc-bus-delay-data-2015.xlsx",sheet_name = None), ignore_index = True)
yr_2016 = pd.concat(pd.read_excel("ttc-bus-delay-data-2016.xlsx",sheet_name = None), ignore_index = True)
yr_2017 = pd.concat(pd.read_excel("ttc-bus-delay-data-2017.xlsx",sheet_name = None), ignore_index = True)
yr_2018 = pd.concat(pd.read_excel("ttc-bus-delay-data-2018.xlsx",sheet_name = None), ignore_index = True)
yr_2019 = pd.concat(pd.read_excel("ttc-bus-delay-data-2019.xlsx",sheet_name = None), ignore_index = True)
yr_2020 = pd.concat(pd.read_excel("ttc-bus-delay-data-2020.xlsx",sheet_name = None), ignore_index = True)
yr_2021 = pd.concat(pd.read_excel("ttc-bus-delay-data-2021.xlsx",sheet_name = None), ignore_index = True)
yr_2022 = pd.concat(pd.read_excel("ttc-bus-delay-data-2022.xlsx",sheet_name = None), ignore_index = True)
yr_2023 = pd.concat(pd.read_excel("ttc-bus-delay-data-2023.xlsx",sheet_name = None), ignore_index = True)

Converting the excel files to csv for faster loading and processing in jupyter notebooks

In [13]:
yr_2014.to_csv("2014", index = None, header = True)
yr_2015.to_csv("2015", index = None, header = True)
yr_2016.to_csv("2016", index = None, header = True)
yr_2017.to_csv("2017", index = None, header = True)
yr_2018.to_csv("2018", index = None, header = True)
yr_2019.to_csv("2019", index = None, header = True)
yr_2020.to_csv("2020", index = None, header = True)
yr_2021.to_csv("2021", index = None, header = True)
yr_2022.to_csv("2022", index = None, header = True)
yr_2023.to_csv("2023", index = None, header = True)

All the csv files will be read into this notebook, automatically converting them into a dataframe and then merged into one large dataset. 

In [14]:
df_14 = pd.read_csv("2014")
df_15 = pd.read_csv("2015")
df_16 = pd.read_csv("2016")
df_17 = pd.read_csv("2017")
df_18 = pd.read_csv("2018")
df_19 = pd.read_csv("2019")
df_20 = pd.read_csv("2020")
df_21 = pd.read_csv("2021")
df_22 = pd.read_csv("2022")
df_23 = pd.read_csv("2023")

In [18]:
ttc_df = pd.concat([df_14,df_15,df_16,df_17,df_18,df_19,df_20,df_21,df_22,df_23], keys = ["2014","2015","2016","2017","2018","2019","2020","2021","2022","2023"])


In [25]:
ttc_df

Report Date Route      Time        Day                Location  \
2014 0      2014-01-01    95  00:23:00  Wednesday      York Mills station   
     1      2014-01-01   102  00:55:00  Wednesday    Entire run for route   
     2      2014-01-01    54  01:28:00  Wednesday     lawrence and Warden   
     3      2014-01-01   112  01:30:00  Wednesday         Kipling Station   
     4      2014-01-01    24  01:37:00  Wednesday        VP and Ellesmere   
...                ...   ...       ...        ...                     ...   
2023 20235         NaN    11     23:57  Wednesday      DAVISVILLE STATION   
     20236         NaN    11     23:57  Wednesday      DAVISVILLE STATION   
     20237         NaN    37     00:09  Wednesday  ISLINGTON STATION (OUT   
     20238         NaN    32     00:32  Wednesday        EGLINTON STATION   
     20239         NaN    43     01:00  Wednesday   KENNEDY AND BRIGSTOCK   

                         Incident  Min Delay  Min Gap Direction  Vehicle  \
2014 0                 Mechanical       10.0     20.0         E   1734.0   
     1              General Delay       33.0     66.0       b/w   8110.0   
     2                 Mechanical       10.0     20.0        WB   7478.0   
     3         Emergency Services       18.0     36.0         N   8084.0   
     4              Investigation       10.0     20.0         n   7843.0   
...                           ...        ...      ...       ...      ...   
2023 20235  Operations - Operator       25.0     50.0         E   8782.0   
     20236          General Delay       25.0     50.0         S   8782.0   
     20237     Emergency Services       30.0     60.0         S   3343.0   
     20238             Mechanical       15.0     30.0       NaN   1353.0   
     20239             Mechanical       10.0     20.0         S   3118.0   

             Min Delay  Incident ID  Delay  Gap        Date Line Bound  \
2014 0             NaN          NaN    NaN  NaN         NaN  NaN   NaN   
     1             NaN          NaN    NaN  NaN         NaN  NaN   NaN   
     2             NaN          NaN    NaN  NaN         NaN  NaN   NaN   
     3             NaN          NaN    NaN  NaN         NaN  NaN   NaN   
     4             NaN          NaN    NaN  NaN         NaN  NaN   NaN   
...                ...          ...    ...  ...         ...  ...   ...   
2023 20235         NaN          NaN    NaN  NaN  2023-05-31  NaN   NaN   
     20236         NaN          NaN    NaN  NaN  2023-05-31  NaN   NaN   
     20237         NaN          NaN    NaN  NaN  2023-05-31  NaN   NaN   
     20238         NaN          NaN    NaN  NaN  2023-05-31  NaN   NaN   
     20239         NaN          NaN    NaN  NaN  2023-05-31  NaN   NaN   

           Unnamed: 10  
2014 0             NaN  
     1             NaN  
     2             NaN  
     3             NaN  
     4             NaN  
...                ...  
2023 20235         NaN  
     20236         NaN  
     20237         NaN  
     20238         NaN  
     20239         NaN  

[611788 rows x 18 columns]

The dataset has over 600,000 rows. 
Insights gained from the first few rows:
- We see that 'Year' ranges from 2014 to 2023. 
- The 'Time' column has some values in 12 hr format and others in 24 hr format
- The 'Report Date' and 'Date' columns have the same type of values
- The 'Location' and 'Direction' colums have are a mixture of different cases: upper and lower cases

As 'Route', 'Day', 'Location', and 'Incident' seem to be categorical variables, we can find out how many unique values exist for each of these.  

In [26]:
# Route
ttc_df["Route"].drop_duplicates()

2014  0         95
      1        102
      2         54
      3        112
      4         24
              ... 
2023  11765    128
      13578    810
      16108    232
      17058    202
      19728    200
Name: Route, Length: 935, dtype: object

There are 935 unique values in the 'Route' column indicating 935 different routes in the TTC system. This is more than the actual number of routes in the transit system because this list contains inactive/cancelled routes such as 232 and 810.

In [27]:
# Day
ttc_df["Day"].drop_duplicates()

2014  0       Wednesday
      125      Thursday
      588        Friday
      1113     Saturday
      1338       Sunday
      1435       Monday
      1857      Tuesday
Name: Day, dtype: object

There are 7 unique days as expected. The transit system is operative 7 days in a week.

In [28]:
# Location
ttc_df["Location"].drop_duplicates()

2014  0            York Mills station
      1          Entire run for route
      2           lawrence and Warden
      3               Kipling Station
      4              VP and Ellesmere
                        ...          
2023  20204    GLENCAIRN AD DANESBURY
      20211      ATTWELL AND DISCO RD
      20212      GRENOBLE AND VANDOME
      20214     MCCOWAN AND MILNER (S
      20237    ISLINGTON STATION (OUT
Name: Location, Length: 120699, dtype: object

There are apparently 120669 unique locations but this is more than the actual number due to the different cases influencing the uniqueness of the values.

In [29]:
ttc_df["Incident"].drop_duplicates()

2014  0                                Mechanical
      1                             General Delay
      3                        Emergency Services
      4                             Investigation
      11                                Diversion
      20                       Utilized Off Route
      22                      Late Leaving Garage
2019  14555                                Vision
      19419                                   NaN
      24221        Late Leaving Garage - Operator
      24223      Late Leaving Garage - Mechanical
      24247      Late Leaving Garage - Management
      52475          Late Leaving Garage - Vision
2020  15605                            Management
      24860                 Operations - Operator
      24861                              Cleaning
      24867                              Security
      24870                       Collision - TTC
      24887      Road Blocked - NON-TTC Collision
      25037        Road Block - Non-TTC Collision


There seem to be more incident categories than expected. This can be attributed to the different wordings used for incident types of the same nature. 